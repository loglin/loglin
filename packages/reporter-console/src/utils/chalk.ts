import chalk, { Chalk, ForegroundColor, BackgroundColor } from 'chalk'

type ColorCache = Record<string, Chalk>

const colorCache: ColorCache = {}

export function chalkColor(name: string) {
  let color = colorCache[name]
  if (color) {
    return color
  }

  if (name[0] === '#') {
    color = chalk.hex(name)
  } else {
    try {
      color = chalk[name as typeof ForegroundColor] || chalk.keyword(name)
    } catch {
      throw new Error(
        chalk.red(
          `Color "${chalk.bold(
            name
          )}" is not a valid chalk color. Ignoring it.\n`
        )
      )
    }
  }

  colorCache[name] = color
  return color
}

const bgColorCache: ColorCache = {}

export function chalkBgColor(name: string) {
  let color = bgColorCache[name]
  if (color) {
    return color
  }

  if (name[0] === '#') {
    color = chalk.bgHex(name)
  } else {
    try {
      color =
        chalk[
          ('bg' +
            name[0].toUpperCase() +
            name.slice(1)) as typeof BackgroundColor
        ] || chalk.bgKeyword(name)
    } catch {
      throw new Error(
        chalk.red(
          `Background color "${chalk.bold(name)}" is not a valid chalk color.`
        )
      )
    }
  }

  bgColorCache[name] = color
  return color
}
