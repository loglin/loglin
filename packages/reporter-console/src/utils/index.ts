export * from './chalk'
export * from './error'
