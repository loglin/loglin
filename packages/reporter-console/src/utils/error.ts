import { sep } from 'path'
import chalk from 'chalk'

export function parseStack(stack: string) {
  const cwd = process.cwd() + sep

  const lines = stack
    .split('\n')
    .splice(1)
    .map((l) => {
      const line = l.trim().replace('file://', '').replace(cwd, '')

      if (process.env.NODE_ENV === 'test') {
        return line.replace(/\\/g, '/')
      }

      return line
    })

  return lines
}

export function formatError(error: Error) {
  if (!error.stack) {
    return
  }

  const stack =
    '\n\n' +
    parseStack(error.stack)
      .map(
        (line) =>
          '  ' +
          line
            .replace(/^at +/, (m) => chalk.gray(m))
            .replace(/\((.+)\)/, (_, m) => `(${chalk.cyan(m)})`)
      )
      .join('\n') +
    '\n'

  return stack
}
