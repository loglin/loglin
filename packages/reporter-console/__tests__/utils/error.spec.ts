import { parseStack, formatError } from '../../src/utils/error'

describe('parseStack', () => {
  const parsedStack = parseStack(new Error('Error').stack!)

  test('should return array with lines', () => {
    expect(Array.isArray(parsedStack)).toBe(true)
  })
})

describe('formatError', () => {
  test('should return undefined if error has no stack', () => {
    const error = new Error('Error')
    error.stack = undefined

    expect(formatError(error)).toBeUndefined()
  })

  test('should format error', () => {
    const error = new Error('Error')
    error.stack =
      'Error\n' + ['at line 1 (path)', 'at line 2 (path)'].join('\n')
    expect(formatError(error)).toMatchSnapshot()
  })
})
