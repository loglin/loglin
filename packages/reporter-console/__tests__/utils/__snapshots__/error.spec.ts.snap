// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatError should format error 1`] = `
"

  at line 1 (path)
  at line 2 (path)
"
`;
